substratevm/src/com.oracle.svm.core.jdk11/src/com/oracle/svm/core/jdk/Target_java_lang_Module.java

```diff
@@ -0,0 +1,219 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.jdk;
+
+import com.oracle.svm.core.SubstrateUtil;
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.RecomputeFieldValue;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import com.oracle.svm.core.jdk.resources.ResourceStorageEntry;
+
+import javax.lang.model.SourceVersion;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.util.Objects;
+
+@SuppressWarnings("unused")
+@TargetClass(className = "java.lang.Module", onlyWith = JDK11OrLater.class)
+public final class Target_java_lang_Module {
+
+    @SuppressWarnings("static-method")
+    @Substitute
+    public InputStream getResourceAsStream(String name) {
+        ResourceStorageEntry res = Resources.get(name);
+        return res == null ? null : new ByteArrayInputStream(res.getData().get(0));
+    }
+
+    //Checkstyle: allow synchronization
+    @Substitute
+    private static void defineModule0(Module module, boolean isOpen, String version, String location, String[] pns) {
+        if (Objects.isNull(module)) {
+            throw new NullPointerException("Null module object");
+        }
+
+        if (Objects.isNull(module.getName())) {
+            throw new IllegalArgumentException("Module name cannot be null");
+        }
+
+        if (module.getName().equals("java.base")) {
+            if (isOpen) {
+                throw new AssertionError("java.base module cannot be open");
+            }
+
+            for (String pn : pns) {
+                if (!ModuleUtil.isValidPackageName(pn)) {
+                    throw new IllegalArgumentException("Invalid package name: " + pn + " for module: java.base");
+                }
+            }
+
+            if (module.getClassLoader() != null) {
+                throw new IllegalArgumentException("Class loader must be the boot class loader");
+            }
+
+            synchronized (ModuleUtil.moduleLock) {
```

Comment:
Does `defineModule0` have to provide certain guarantees regarding synchronization that require you to have this?

Reply:
Native `defineModule0` method uses mutexes when accessing module/package tables. If we do not synchronize, we might end up allowing the same package to be defined twice to the same classloader for example.